### PR TITLE
Increase training paradigm flexibility 

### PR DESCRIPTION
**Pull request by mo-sameh**
In this pull request, I'm suggesting some minor changes that we used for Alphadia project that we've found beneficial and believe could also be useful for some alphapeptdeep users. Since the training process are inherited directly from the model interface rather than through composition. Consequently, making modifications to the training process using the existing interfaces posed a greater challenge. The focus of this PR is to provide more access to the training paradigm from the model interface without changing the architecture and more importantly ensuring compatibility with existing pipelines . 

Here are the proposed changes:

1) Epoch and batch callbacks:
      These callbacks are triggered after each epoch and batch during training. They pass information such as the epoch/batch 
number and training loss, enabling monitoring and reporting of the loss between training epochs/batches. However, the key feature of these callbacks is the ability to return a "continue training" boolean value that could be used to stop the training. We utilized this feature to implement early stopping in the Alphadia fine-tuning module.

2) Learning rate scheduler:
      This change allows users to  alter the learning rate scheduler as long as it conforms to the 'LR_SchedulerInterface'. One issue here is that the scheduler is currently initialized within the class. So to maintain backward compatibility the interface takes a class as an argument and initialize it in-ModelInterface, instead of using an object of the LR_scheduler.
      
- Model Manager save:
      Included the charged model in the save models function for completeness.

Overall, we believe these proposed solutions should not disrupt any existing pipelines but rather enhance the flexibility of the training loops particularly for users who wish to extend the existing training framework. If you have any concerns about these changes or have alternative suggestions for achieving the same goal, please let me know.

### CODE REVIEW

File: peptdeep/model/model_interface.py [hunk 1]
```diff
@@ -125,6 +223,9 @@ def __init__(self,
         self.set_device(device)
         self.fixed_sequence_len = fixed_sequence_len
         self.min_pred_value = min_pred_value
+        self.epoch_callback = default_epoch_callback
```
Comment (mschwoer) [inline peptdeep/model/model_interface.py:226]: have you considered handling the callbacks also within a dedicated class (like `CallbackHandler`)? The default implementation simply does nothing, and if you want a dedicated callback, you would overwrite it.
You would have even more encapsulation then (the `default_epoch/batch_callback` methods are a bit dangling here)

File: peptdeep/model/model_interface.py [hunk 2]
```diff
@@ -185,6 +286,68 @@ def target_column_to_train(self)->str:
...
+        func : callable
+            The callback function.
+        """
+        # Check that the function is callable and has the right signature
+        if not callable(func):
+            raise ValueError("The callback function must be callable")
+        # Make sure that the only arguments with default values are the ones that are expected
+        for arg in inspect.signature(func).parameters:
+            if arg not in ['epoch', 'epoch_loss'] and isinstance(inspect.signature(func).parameters[arg].default, type(inspect._empty)):
+                raise ValueError("The callback function must have only two arguments without default values: epoch and epoch_loss")
```
Comment (mschwoer) [inline peptdeep/model/model_interface.py:308]: There's quite similar code in `set_batch_callback`.
What about introducing a helper method `_check_callback_impl(arg_list: List[str])` (or the like) ?

File: peptdeep/model/model_interface.py [hunk 3]
```diff
@@ -32,56 +32,154 @@
...
+    pass
+
+class LR_SchedulerInterface(object):
+    def __init__(self, optimizer:torch.optim.Optimizer, **kwargs):
+        raise NotImplementedError
+
+    def step(self, epoch:int, loss:float):
+        """
+        This method must be implemented in the sub-class. It will be called to get the learning rate for the next epoch. 
+        While the one we are using here does not need the loss value, this is left incase of using something like the ReduceLROnPlateau scheduler.
```
Comment (mschwoer) [inline peptdeep/model/model_interface.py:71]: (nit) this line describes an actual implementation, of which the abstract class does not know of :-)

File: peptdeep/model/model_interface.py [hunk 4]
```diff
@@ -32,56 +32,154 @@
...
+        Returns
+        -------
+        float
+            The last learning rate.
+        """
+        raise NotImplementedError
+    
+class WarmupLR_Scheduler(LR_SchedulerInterface):
+    """
+    A Lr scheduler that includes a warmup phase and then a cosine annealing phase.
```
Comment (mschwoer) [inline peptdeep/model/model_interface.py:95]: (nit) "A learning rate scheduler .."

File: peptdeep/model/model_interface.py [hunk 5]
```diff
@@ -32,56 +32,154 @@
...
+
+
+    # `transformers.optimization.get_cosine_schedule_with_warmup` will import tensorflow,
+    # resulting in some package version issues.
+    # Here we copy the code from transformers.optimization
+    def _get_cosine_schedule_with_warmup_lr_lambda(self,
+        current_step: int, *, num_warmup_steps: int, num_training_steps: int, num_cycles: float
+    ):
+        if current_step < num_warmup_steps:
+            return float(current_step) / float(max(1, num_warmup_steps))
```
Comment (mschwoer) [inline peptdeep/model/model_interface.py:143]: the initial code had `return float(current_step+1) ..` here .. was the `+1` removed by accident?

File: peptdeep/model/model_interface.py [hunk 6]
```diff
@@ -298,14 +461,22 @@ def train_with_warmup(self,
                     batch_size, verbose_each_epoch,
                     **kwargs
                 )
+                result = self.epoch_callback(epoch, np.mean(batch_cost))
+                if result == False:
```
Comment (mschwoer) [inline peptdeep/model/model_interface.py:465]: just out of interest: is this idiom ("return false -> stop") common for those callbacks?
Alternatively one could base this on exceptions..

File: peptdeep/model/model_interface.py [hunk 7]
```diff
@@ -32,56 +32,154 @@
...
-    return LambdaLR(optimizer, lr_lambda, last_epoch)
+    def __init__(self,
+        optimizer:torch.optim.Optimizer,
+        num_warmup_steps:int,
+        num_training_steps:int,
+        num_cycles:float=0.5,
+        last_epoch:int=-1
+    ):
+        self.optimizer = optimizer
+        self.lamda_lr = self.get_cosine_schedule_with_warmup(
```
Comment (mschwoer) [inline peptdeep/model/model_interface.py:106]: (nit) typo: `lambda`

File: peptdeep/model/model_interface.py (hunk 2 above)
Comment (mschwoer) [reply peptdeep/model/model_interface.py:308]: another thought: if you introduced a `CallbackHandler` as proposed above, this would be as simple as a check for `isSubclass`

File: peptdeep/model/model_interface.py (hunk 5 above)
Comment (jalew188) [reply peptdeep/model/model_interface.py:143]: +1 must be kept, otherwise lr=0 will be generated in the first epoch, resulted in training nothing.

File: peptdeep/model/model_interface.py (hunk 6 above)
Comment (GeorgWa) [reply peptdeep/model/model_interface.py:465]: We discussed this a bit. 
I think an using an Exception here does not match the idea of an exception as an error. 

It would also require a `try ... except` statement to catch this case. IMHO Using exceptions for expected control flow is somewhat of an antipattern as this builds hidden dependencies or contracts between modules. Instead of using the regular synchronous control flow we rely on the exception being passed up the stack. Therefore, a higher level module needs to know that it has to except the `TrainingFinishedException` anytime at runtime alltough this is not part of the function and class signatures of the submodules.

Some languages such as Rust apply this even to error handling, returning an `Option` to indicate the possibility of an error while maintaining control flow. https://doc.rust-lang.org/std/option/

File: peptdeep/model/model_interface.py (hunk 1 above)
Comment (GeorgWa) [reply peptdeep/model/model_interface.py:226]: Can you explain this a bit more? I don't understand it.

File: peptdeep/model/model_interface.py (hunk 5 above)
Comment (jalew188) [reply peptdeep/model/model_interface.py:143]: Yes, this is a bug fix sometime ago

Comment (jalew188) [reply peptdeep/model/model_interface.py:143]: @mschwoer This is a good case showing why a small PR is necessary for review as small PRs can keep the code up to date. We shall do "git pull/merge" before making changes. 

Small issues may be ignored during review in a big PR.
